### PR TITLE
Feature/add some implementation

### PR DESCRIPTION
**Pull request by smoser**
This is some initial implementation as moved over from elsewhere.

I'm interested in feedback on the DiskImpl (name) and convention of implementation and such.

I do have  some tests to move over still, and what is here probably isn't even complete, but would like some feedback.

### CODE REVIEW

**Comment by rchamarthy** [comment]
@smoser rebase your PR latest origin/master and the code coverage should work.

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+issues:
+  exclude-rules:
+    - linters: ["gocritic"]
+      text: "ifElseChain:"
+    - linters: ["gosec"]
+      text: "G204:"
+      path: util.go
+    # 1 is not a magic number. 3 is a magic number.
+    - linters: ["gomnd"]
```
Comment (rchamarthy) [inline .golangci.yml]: I am actually debating if we should just disable this linter. Do you see value? If so lets use the // nolint method to suppress where applicable.

File: .golangci.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  wsl:
+    allow-cuddle-declarations: true
+
+issues:
+  exclude-rules:
+    - linters: ["gocritic"]
+      text: "ifElseChain:"
+    - linters: ["gosec"]
+      text: "G204:"
+      path: util.go
```
Comment (rchamarthy) [inline .golangci.yml]: If it is just for util.go its better to use //nolint:gosec on the offending line, that way while reading code its apparent that we are suppressing that and we know the context/reason.

File: .golangci.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+linters-settings:
+  wsl:
+    allow-cuddle-declarations: true
+
+issues:
+  exclude-rules:
+    - linters: ["gocritic"]
+      text: "ifElseChain:"
```
Comment (rchamarthy) [inline .golangci.yml]: I would also keep this in general and exclude when required. But i will leave it up to you.

File: disk.go [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/apex/log"
```
Comment (rchamarthy) [inline disk.go]: Using logging in this context is probably the same as using fmt.Printf as we wont log to files. Can we do without logging for library.

File: util.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/apex/log"
+	"github.com/pkg/errors"
```
Comment (rchamarthy) [inline util.go]: we will also need corresponding go.mod, go.sum changes.

File: virt.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+				systemVirtType = virtKvm
+			case "lxc":
+				systemVirtType = virtLxc
+			case "":
+				systemVirtType = virtUnknown
+
+				log.Debugf("Unknown virt type: %s/%s", string(out), string(stderr))
+			}
+		} else {
+			log.Warnf("Failed to read virt type [%d]: %s/%s",
```
Comment (rchamarthy) [inline virt.go]: We should probably just return this as an error.

File: .golangci.yml (hunk 1 above)
Comment (smoser) [reply .golangci.yml]: I dont have a strong feeling on it. It seems reasonable to not have hard coded numbers like 17.
But I just do not think that '1' is a hard coded magic number.

I'd suggest we leave this as it is. If its annoying we can get rid of it later.  At this point 0 and 1 are allowed.

File: .golangci.yml (hunk 2 above)
Comment (smoser) [reply .golangci.yml]: yeah, i guess you're right here.  I really do nnont like '//nolint' lines generally, but this really is one line t hat we're saying is OK for this.

File: disk.go (hunk 4 above)
Comment (smoser) [reply disk.go]: I'll drop the logging for now.

File: util.go (hunk 5 above)
Comment (smoser) [reply util.go]: ack

**Pull request merged**